src/common/tracer.h

```diff
@@ -0,0 +1,80 @@
+#ifndef TRACER_H_
+#define TRACER_H_
+
+#define SIGNED_RIGHT_SHIFT_IS 1
+#define ARITHMETIC_RIGHT_SHIFT 1
+
+#include<cstring>
+#include <arpa/inet.h>
+#include <yaml-cpp/yaml.h>
+#include<atomic>
+
+#include <jaegertracing/Tracer.h>
+#include <jaegertracing/net/IPAddress.h>
+#include <jaegertracing/net/Socket.h>
+
+
+typedef std::unique_ptr<opentracing::Span> Span;
+
+struct Jaeger_Tracer{
+    Jaeger_Tracer(){}
+    ~Jaeger_Tracer();
+  std::shared_ptr<opentracing::v3::Tracer> tracer = NULL;
+};
+
+//will be used throughout ceph to create spans
+extern Jaeger_Tracer tracer;
+extern std::atomic<bool> jaeger_initialized;
+
+
+static inline void init_tracer(const char* tracerName,const char* filePath){
+      if(jaeger_initialized) return;
+
+      try{
+          auto yaml = YAML::LoadFile(filePath);
+          auto configuration = jaegertracing::Config::parse(yaml);
+
+          jaegertracing::net::Socket socket;
+          socket.open(AF_INET, SOCK_STREAM);
+          const std::string serverURL = configuration.sampler().samplingServerURL();
+          socket.connect(serverURL); // this is used to check if the tracer is able to connect with server successfully
+
+          tracer.tracer = jaegertracing::Tracer::make(
+          tracerName,
+          configuration,
+          jaegertracing::logging::consoleLogger());
+      }catch(...) {return;}
+      opentracing::Tracer::InitGlobal(
+      std::static_pointer_cast<opentracing::Tracer>(tracer.tracer));
+      jaeger_initialized = true;
+  }
+
+static inline Span new_span(const char* span_name){
+      Span span=opentracing::Tracer::Global()->StartSpan(span_name);
+      return span;
+  }
+static inline Span child_span(const char* span_name, const Span& parent_span){
```

Comment:
passing a unique pointer as reference is usually an indications that something is wrong here.
if you don't want to pass ownership, pass a a bare pointer, if you want to pass ownership pass unique pointer by value

Reply:
I just want a const reference here so that the child_span can use this to refer
https://github.com/jaegertracing/jaeger-client-cpp/blob/master/examples/App.cpp

Reply:
then just pass a const bare pointer. there is no ownership passing here

Reply:
this cant be done because the parent_span is is a nullptr as default value so  `&parent_span` will cause crash, and this will happen when we `WITH_JAEGER` flag off. So I think it is safe pass them as reference.

Reply:
i don't understand how passing a bare pointer is causing a crash?
IMO, best practice around unique_ptr is to have it as a local variable in the function where you allocate it, and pass it as a bare pointer to all other functions.
if it is nullptr by default, you can check if it is null before you use it.
i don't see how pointer ownership needs to be handled by anyone else than the first function?

Reply:
1.Reference passing does change owernship of smart pointers.
2.the checking method will add a extra line of code which can eliminated if I use my reference passing mechanism.

I m not passing it as std::move which causes the ownership to change.

https://stackoverflow.com/questions/17368644/passing-smart-pointers-by-reference#:~:text=It's%20ok%20to%20pass%20a,corruption%20if%20you%20did%20that.

Reply:
1. right, but if you don't want to change ownership you don't need a smart pointer in the API to begin with
2. i dont get that, if the pointer is null you need to check it regardless of whether it is a unique pointer or not

the stackoverflow comment discussion says:
* don't pass them by-value
* ok to pass by reference (unless you pass into a reference data member)

If something is OK it means you CAN do that, not that you SHOULD
I don't see any point in using them in the API at all. except for introducing unnecessary confusion as for why do you use them. when someone sees a unique_ptr used in the API, they assume that there is some ownership change in the functions, which is not the case here.

Reply:
@yu
I could have done to things here either make this 
```
if(parent_span)
  child("name",parent_span)` 
```
everywhere, which is just makes the source code a little uglier and long too, doing function call according to me is much clean :)
Second yes passing reference to of unique_ptr is bad, but only if we are storing it somewhere like either calling a constructor with this reference or using it initialize a unique_ptr, but in no way I m storing this reference, this reference is used by jaeger internal library call which I think only extracts the id of the parent_span to make child--parent relationship into its backend, also the use of `const` confirms this that in no way there will be a change of ownership as std::move also gets ruled beacuse of it.

The main strong reason I m using it because I saw that jaeger-client-cpp itself uses this to in its example code. https://github.com/jaegertracing/jaeger-client-cpp/blob/master/examples/App.cpp the example

Reply:
i dont understand why you need to check the pointer before passing it in.
just pass nullptr.
you need to check for null only before you dereference the pointer.

on top of that, you need to check if a unique_ptr is null before dereferencing it as well, so i don't understand why it makes the code uglier?

Reply:
Yeah sorry my bad, I was speaking without testing it, there is no need for checking null condition while calling. But dont you think that this wont affect anything, because both will be working fine :)

Reply:
functionally both are working the same.
however, adding unnecessary complexity to APIs is always bad.
anyone reading the code will have to dig into it and figure out why you pass the unique_ptr, and whether you have ownership change to the pointer - this is unnecessary mental burden